### ISSUE TITLE
Implement chunk/extent birth rank as a higher fidelity packing mechanism.

### ISSUE THREAD

**Issue by jasone** [jemalloc/jemalloc#147]
jemalloc prefers the lowest available address when allocating a new region, because this stable ordering tends to pack allocations toward the bottom of memory and allow higher memory to drain of allocations that would otherwise cause chunk/page-level fragmentation.  However, the key here is that addresses provide a stable ordering, rather than lower addresses being inherently superior.

Experiment with storing a birth serial number in every arena chunk, and in every page run, so that rather than preferring lower addresses, jemalloc prefers lower birth serial numbers.  The effect should be similar, but perhaps superior, since the ultimate goal is to pack the oldest chunks/runs most tightly.  Two things may keep this from being a net win:
1. Birth serial numbers require space to store, to the tune of ~0.2% of memory.
2. Arena chunks may not completely drain as often, which will leave more extant arena chunks, and therefore more arena chunk headers.

**Comment by jasone** [comment]
Now that page runs for small allocations store their metadata in chunk headers, a birth serial number per run induces unacceptable overhead.  However, chunk birth serial numbers are still feasible.

**Comment by jasone** [comment]
Late in the 4.3.[01] release cycle I spent some time looking at virtual memory fragmentation issues due to bugs, and although those bugs are fixed now, I came away realizing just how poorly the "pack to low addresses" heuristic can behave if ```mmap(2)``` tends to place subsequent mappings lower (rather than higher) in memory.  I originally implemented the placement policies in the context of FreeBSD where memory grew upward (whether due to ```sbrk(2)``` or ```mmap(2)``` I don't remember), and graphs showed really good behavior.  However, at least on the Linux systems I've examined recently this is a non-trivial flaw that I feel strongly about fixing ASAP.

In the context of pre-5.0.0 (chunks), this can all be solved at the per chunk level by tracking chunk creation order using serial numbers; there is no need to change anything about how runs and small allocation placement works except to incorporate their enclosing chunks' serial numbers into placement decisions.  In the context of 5.0.0, we need to track extent creation order rather than chunk creation order, but otherwise the principles are the same.  In both cases, we need to take a bit of care to consistently merge serial numbers when merging adjacent mappings (e.g. preserve the lowest serial number).  The result will be that we achieve stable placement ordering regardless of whether memory grows upward or downward.

**Comment by ronawho** [comment]
@jasone upgrading to 4.4.0 resulted in a minor but non-trivial performance regression (~5% slowdown) for an allocation-heavy benchmark. I've bisected down to this patch (5c77af98b16a0f5b15bc807f2b323a91fe2a048b.) Personally, I'm not surprised that this could hurt performance a little, particularly since it adds a few atomic operations.

I'm fine living with the regression, but I wanted to check that a performance loss from this isn't surprising to you. If it is surprising, I can open up a new issue and try to come up with a relatively simple reproducer.

**Comment by jasone** [comment]
@ronawho, I've some minor throughput regressions in some cases, but in every case I've dug into, it has been due to reduced fragmentation allowing page runs to completely drain more often, resulting in more chunk and run allocation/deallocation (as well as reduced fragmentation).  If your benchmark is also seeing a higher newruns rate, then this is a good thing overall.  If instead you determine that e.g. atomic operations are causing a bottleneck, please let me know so we can fix it.

Funny related story: there were some pretty terrible chunk fragmentation issues in early 3.x releases (pre-3.2.0), and some people delayed for a long time before updating because they didn't care about the virtual memory issues, but their benchmarks measured a throughput regression they didn't like.

**Comment by ronawho** [comment]
We typically build with `--disable-stats`. Enabling stats makes the performance delta between 4.3.1 and 4.4.0 smaller (4.4 is only ~2% slower with stats enabled):

| config                | time (s) |
| --------------------- | -------- |
| 4.3.1 --disable-stats |     3.50 |
| 4.4.0 --disable-stats |     3.68 |
| 4.3.1 w/ stats        |     3.75 |
| 4.4.0 w/ stats        |     3.82 |

note that the timings are really stable/consistent


The stats report between 4.3.1. and 4.4.0 look very, very similar. The number of newruns is the same, but the number of reruns is ever so slightly lower for 4.4.0, though only for size class 32. That said, it's only 8 less (1,099,673 vs. 1,099,665):

| config | size ind | allocated |   nmalloc |   ndalloc | nrequests | curregs | curruns | regs | pgs |  util |   nfills | nflushes | newruns |  reruns |
| ------ | -------- | --------- | --------- | --------- | --------- | ------- | ------- | ---- | --- | ----- | -------- | -------- | ------- | ------- |
| 4.3.1  |   32   2 |     51456 | 490385056 | 490383448 | 542477781 |    1608 |     146 |  128 |   1 | 0.086 | 18949610 |  4916578 | 3493609 | 1099673 |
| 4.4.0  |   32   2 |     51456 | 490385056 | 490383448 | 542477781 |    1608 |     146 |  128 |   1 | 0.086 | 18949610 |  4916578 | 3493609 | 1099665 |


I'm not sure what exactly to take away from all that though..

**Comment by ronawho** [comment]
And FWIW in some sense this is a "lame" benchmark that does almost nothing beyond allocate and deallocate a lot of short lived objects in parallel, so maybe it's not worth digging into at all since it's not representative of real codes

The benchmark is actually [Chapel's binary-trees submission](http://benchmarksgame.alioth.debian.org/u64q/program.php?test=binarytrees&lang=chapel&id=1) for the computer language benchmarks game

Looking at a reference [C version](http://benchmarksgame.alioth.debian.org/u64q/program.php?test=binarytrees&lang=gcc&id=5) I see similar perf loss between 4.3.1 and 4.4.0 (although again it's a really small difference)

| config                | time (s) |
| --------------------- | -------- |
| 4.3.1 w/ stats        |     3.42 |
| 4.4.0 w/ stats        |     3.51 |
| system (just for fun) |     7.02 |

**Comment by jasone** [comment]
I spent a while with the C version of binary-trees, and ended up with some conflicting results.  At some point earlier today I measured a slowdown similar to what you reported, and noticed that more purging was happening for 4.4.0.  I tried several tweaks, e.g. ```narenas:1``` (nearly identical results between versions) and ```purge:decay``` (4.4.0 slightly faster).  My tentative conclusion was that the slowdown with default options was an artifact of layout that triggered more purging, but when I went to reproduce the original result this evening, I instead measured a consistent 5-7% *speedup* with 4.4.0 vs. 4.3.1.  Maybe it's timing-sensitive, or maybe the kernel on my dev system entered a distinctly different mode part way through the day.  I don't think atomic operations are a major factor though; for this benchmark I count only an additional 117 atomic adds.

**Comment by ronawho** [comment]
Thanks for digging into this, really appreciate it.

I got distracted with other work for a while, but I was looking into this again today. At the time I reported this, I think I was running on a Cray compute node (really stripped down OS with almost no other services/processes running) so I was getting really stable timings. On our main dev boxes I also see less reliable timings, especially with the C version. The timings I reported were with purge:decay (we use it by default), so it's a little curious that you were seeing 4.4.0 running faster than 4.3.1 with purge:decay enabled.

If I bump the problem size up from 20 to 24 or 26 I see relatively consistent timings. 4.4.0 is still around 4-5% slower, but I think you're right about the slowdown being caused by more purging, especially since I'm not seeing any regressions on real benchmarks/codes.

Thanks again!
